internal/incident/incident.go

```diff
@@ -173,13 +182,31 @@ func (i *Incident) ProcessEvent(ctx context.Context, ev *event.Event) error {
 		}
 	case event.TypeAcknowledgementSet:
 		if err := i.processAcknowledgementEvent(ctx, tx, ev); err != nil {
+			if errors.Is(err, errSuperfluousAckEvent) {
+				// That ack error type indicates that the acknowledgement author was already a manager, thus
+				// we can safely ignore that event and return without even committing the DB transaction.
+				return nil
+			}
+
 			return err
 		}
 	}
 
-	notifications, err := i.addPendingNotifications(ctx, tx, ev, i.getRecipientsChannel(ev.Time))
-	if err != nil {
-		return err
+	var notifications []*NotificationEntry
+	switch {
+	// If the current incident is unmuted, we want to always generate/send notifications, otherwise, if the current
+	// object is unmuted, we also want to generate/send Incident/Object unmuted notifications with the ongoing event.
+	case !i.isMuted || !i.Object.IsMuted():
+		fallthrough
+		// On the other hand, if the current incident and its object are currently muted, and we do not want to
+		// generate any notifications if the type of the ongoing event is either Downtime/Flapping-Start or ACK-set,
+		// but simply ignore them silently. For all other remaining event types, such as state changes etc., we are
+		// going to record "suppressed" notifications in the database, though not actually send them.
+	case ev.Type != event.TypeDowntimeStart && ev.Type != event.TypeAcknowledgementSet && ev.Type != event.TypeFlappingStart:
+		notifications, err = i.generateNotifications(ctx, tx, ev, i.getRecipientsChannel(ev.Time))
+		if err != nil {
+			return err
+		}
 	}
```

Comment:
So, if I understand correctly, `i.isMuted != i.Object.IsMuted()` if and only if the current event mutes or unmutes the object. Thus, the first `case` triggers notifications if the object is being muted/unmuted by the current event, or if the object is not muted (and the current event doesn't change that).

The second `case` then only affects "nested" muting start events (we don't track nested muting, this is just saying that another "start muting" event happens while already muted). I presume the asymmetry that it checks for the start but not for the end events is because end events are only generated when all muting reasons are gone, i.e. there won't be a downtime end event while still being acknowledged?

Did I get that correctly so far?

Reply:
> Did I get that correctly so far?

Yes!

Reply:
Was that a general question of clarification or do you want me to extend the existing comments?

Reply:
So far only general questions. I'm trying to understand it so that I can see if I can come up with a way to simplify this.

> because end events are only generated when all muting reasons are gone

But now I'm confused about that part after looking at the code again:

https://github.com/Icinga/icinga-notifications/blob/a07add3c95d5a8a63d6885a24dff1f799de2e324/internal/icinga2/client.go#L347-L353

This looks like a flapping stop event would still be emitted even while still muted by another reason (code for ack clear and downtime end looks very similar), it just doesn't say unmute but keeps the current muting state. Is there any other place in the code that would stop the processing of these events?

However, emitting and processing all (including nested) {downtime,flapping,ack} {start,end} events sounds like a good idea, after all, we want to be able to notify for them in the future.

What would happen if the check here was something like this:

```go
if !i.isMuted || !i.Object.IsMuted() || ev.Type == event.TypeState {
	generateNotifications()
}
```

Shouldn't this behave very similar? I.e. notifications outside of any muting are sent as well as for the events actually muting oder unmuting. While being muted, no notifications are sent and only state events generate suppressed history events.

What would happen if `generateNotifications()` was just called here unconditionally and internally decides if a notification is sent or suppressed (just like it already does in the current state of the PR)? Shouldn't this result in a history like the following?

- Ack set (sent)
- Downtime start (suppressed, because already muted by ack)
- Downtime start (suppressed, because already muted)
- State changed / ack cleared (both suppressed, not entirely sure about the exact order, but unless sticky, the ack will be cleared due to the state change, but still in downtime, so still suppressed)
- Downtime end (suppressed, still in the other downtime)
- Downtime end (sent, because this removed the last remaining muting reason)

That would look perfectly fine for me now. And it might even be a good starting point for allowing sending some of these (for now) suppressed notifications.

Reply:
> Is there any other place in the code that would stop the processing of these events?

Why would you want to stop processing these? These are regular Icinga 2 events and not virtually generated ones, and as such they should exist somewhere in the database (in the `event` table) even if we aren't going to trigger notifications for them.

> What would happen if generateNotifications() was just called here unconditionally and internally decides if a notification is sent or suppressed (just like it already does in the current state of the PR)? Shouldn't this result in a history like the following?

In its initial version of this PR, that was exactly the case, and I did asked you explicitly about this (how we should handle two consecutive events), and you told me not even to generate a suppressed notification for these events, but rather to silently ignore them if the object was already muted by previous events. Yet, you are suggesting to restore this behaviour?

Reply:
> > Is there any other place in the code that would stop the processing of these events?
> 
> Why would you want to stop processing these? These are regular Icinga 2 events and not virtually generated ones, and as such they should exist somewhere in the database (in the `event` table) even if we aren't going to trigger notifications for them.

I didn't want to suggest stopping to process these, but I was under the assumption that this was the case and the reason for why `case ev.Type != event.TypeDowntimeStart && ev.Type != event.TypeAcknowledgementSet && ev.Type != event.TypeFlappingStart` only checks for the start but not the end events.

> > What would happen if generateNotifications() was just called here unconditionally and internally decides if a notification is sent or suppressed (just like it already does in the current state of the PR)? Shouldn't this result in a history like the following?
> 
> In its initial version of this PR, that was exactly the case, and I did asked you explicitly about this (how we should handle two consecutive events), and you told me not even to generate a suppressed notification for these events, but rather to silently ignore them if the object was already muted by previous events. Yet, you are suggesting to restore this behaviour?

Sounds like there was a misunderstanding. I was under the impression that these would be generic mute/unmute events. Then, processing a second mute event wouldn't do something useful and I wouldn't have wanted to introduce another "mute reason changed" event. However, when those events still have distinct and useful types, I don't see a reason why they shouldn't be processed.